### ISSUE TITLE
[Bug]: Logprob values are affected by sampling parameters and are incompatible with OpenAI API

### ISSUE THREAD

**Issue by DIYer22** [vllm-project/vllm#9453]
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
root@vllm-cpu:/workspace# python3 collect_env.py 
Collecting environment information...
INFO 08-20 07:37:37 importing.py:10] Triton not installed; certain GPU-related functions will be not be available.
PyTorch version: 2.4.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.30.2
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul 29 2024, 16:56:48) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.10.25-nvidia-gpu-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 57 bits virtual
Byte Order:                      Little Endian
CPU(s):                          17
On-line CPU(s) list:             0-16
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Gold 6330 CPU @ 2.00GHz
CPU family:                      6
Model:                           106
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s):                       17
Stepping:                        6
BogoMIPS:                        4000.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves wbnoinvd arat avx512vbmi umip pku avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid md_clear arch_capabilities
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       544 KiB (17 instances)
L1i cache:                       544 KiB (17 instances)
L2 cache:                        68 MiB (17 instances)
L3 cache:                        272 MiB (17 instances)
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] intel_extension_for_pytorch==2.4.0+gitfbaa4bc
[pip3] numpy==1.26.4
[pip3] pyzmq==26.1.0
[pip3] torch==2.4.0+cpu
[pip3] torchvision==0.19.0+cpu
[pip3] transformers==4.44.0
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.5.4@3f674a49b5033a6ed778ab960e86e03cfa64aa1f
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
Could not collect
```

</details>

### 🐛 Describe the bug

The logprob in vLLM is not the raw probability of the standard LLM loss but is influenced by the sampling parameters. On the other hand, OpenAI returns the raw probability, meaning that no matter how the sampling parameters are set, the logprob of the next token under the same context remains unchanged in OpenAI.

I would like vLLM's logprob to be consistent with OpenAI's behavior because logprob should reflect the ideal probability of a token being sampled, independent of the sampling parameters. Additionally, when setting `top_p: 0.0001`, only the first token in `top_logprobs` is correct, while the subsequent tokens are fixed high-frequency tokens (such as `"`, `!`) instead of the current highest-probability tokens. The bug can be reproduced as follows:

```bash
curl -X POST "http://39.105.21.95:12481/v1/chat/completions" \
     -H "Content-Type: application/json" \
     -d '{
           "model": "meta-llama/Meta-Llama-3-8B-Instruct",
           "top_p": 0.0001,
           "stream": false, "max_tokens":5, "logprobs": true, "top_logprobs": 3,
           "messages": [
             {
               "role": "user",
               "content": "just generate a random two-digit integer , no other words"
             }
           ]
         }'

{'id': 'chat-6f6322c5dfab404782adf729201a024c',
 'object': 'chat.completion',
 'created': 1729071624,
 'model': 'meta-llama/Meta-Llama-3-8B-Instruct',
 'choices': [{'index': 0,
   'message': {'role': 'assistant', 'content': '43', 'tool_calls': []},
   'logprobs': {'content': [{'token': '43',
      'logprob': 0.0,
      'bytes': [52, 51],
      'top_logprobs': [{'token': '43', 'logprob': 0.0, 'bytes': [52, 51]},
       {'token': '"', 'logprob': -9999.0, 'bytes': [34]},
       {'token': '!', 'logprob': -9999.0, 'bytes': [33]}]},
     {'token': '',
      'logprob': 0.0,
      'bytes': [],
      'top_logprobs': [{'token': '', 'logprob': 0.0, 'bytes': []},
       {'token': '"', 'logprob': -9999.0, 'bytes': [34]},
       {'token': '!', 'logprob': -9999.0, 'bytes': [33]}]}]},
   'finish_reason': 'stop',
   'stop_reason': None}],
 'usage': {'prompt_tokens': 21, 'total_tokens': 23, 'completion_tokens': 2},
 'prompt_logprobs': None}
```

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by ctjian** [comment]
I have the same issue

**Comment by tqfang** [comment]
same

**Comment by DIYer22** [comment]
I wrote a python function that restores the sampling probability of vLLM to OpenAI probability (top_p=1, temperature=1) using sampling parameters.

It's only an approximate restoration and can be used temporarily, and it still can't solve the issue where when setting top_p: 0.0001, only the first token in top_logprobs is correct.
```python
import numpy as np

def prob_sample_to_recover(prob_sample, temp, top_p=1, hidden_vocab_num=None):
    prob_sample = prob_sample * top_p
    top_logprobs = len(prob_sample)

    if hidden_vocab_num is None:
        hidden_vocab_num = 1 if (1 - prob_sample.sum() > 0.001) else 0

    if hidden_vocab_num > 0:
        hidden_probs = [max(1 - prob_sample.sum(), 0) / hidden_vocab_num] * hidden_vocab_num
        prob_sample = np.concatenate([prob_sample, hidden_probs])

    prob_recover_unnorm = prob_sample ** temp
    prob_recover = prob_recover_unnorm / prob_recover_unnorm.sum()

    return prob_recover[:top_logprobs]

# Test case

# Sampling parameters
temp = 0.5
top_p = 0.9
top_logprobs = 3

# Original logits
logits = np.array([1, 0, -1, -2,])

# Compute prob_sample using softmax with temperature temp
logits_temp = logits / temp
exp_logits_temp = np.exp(logits_temp)
prob_sample = exp_logits_temp / exp_logits_temp.sum()

prob_sample /= top_p
prob_sample = prob_sample[:top_logprobs]  # Same to vllm's e**logprob

prob_recover = prob_sample_to_recover(prob_sample, temp, top_p)

print(" prob_sample:", prob_sample)
print("prob_recover:", prob_recover)

# Ground truth probabilities
prob_gt = np.exp(logits) / np.exp(logits).sum()
print("     prob_gt:", prob_gt)
```

**Comment by robertgshaw2-neuralmagic** [comment]
@DIYer22 - thanks, Im currently looking at fixing this in VLLM V1. Do you have a reference anywhere to what the OpenAI spec is for logprobs? I could not find anything in their official docs about whether the logprobs are pre- or post- logits processing.

**Comment by DIYer22** [comment]
@robertgshaw2-neuralmagic  Great!

OpenAI does not disclose specific details about its internal implementation, but the following behavior has been observed through its API:

No matter how the sampling parameters are set, the log probability of the next token under the same context remains unchanged in OpenAI.

This means that regardless of how the temperature, top_p are set, the top_logprobs under the same context will always be the same as when temperature=1 and top_p=1.

Additionally, the log probability of prompt_logprobs is also affected by the sampling parameters and needs to be fixed.

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks @DIYer22. This feedback is very helpful.

This is how we have it implemented for the upcoming VLLM V1 release. The PR is WIP: https://github.com/vllm-project/vllm/pull/9880

Will look into updating for current main.

**Comment by gx-ai-architect** [comment]
@robertgshaw2-redhat @wonder if there are PRs aimed to solve this issue?

The feature to obtain logprobs from different temperature is important for some MCMC algorithms to enable inference scaling
